File: src/time_mcp_server/notify_hook.py
"""Claude Code UserPromptSubmit hook: inject notifications for expired timers
and fired alarms into the session.

Configure in ``~/.claude/settings.json``:

    {
      "hooks": {
        "UserPromptSubmit": [
          {
            "command": "C:/Users/<you>/.venvs/time-mcp/Scripts/python.exe -m time_mcp_server.notify_hook"
          }
        ]
      }
    }

The hook drains stdin (the prompt JSON Claude Code sends), reads
``~/.time-mcp/state.json``, and emits one ``additionalContext`` payload for
each timer that has expired or alarm that has fired without being notified.
The ``notified_at`` field is then set on each, so the next run is silent
unless something new fires.

Designed to be cheap: imports only ``time_mcp_server.state`` (stdlib
underneath), no FastMCP / pydantic / dateparser load at hook startup.
"""

from __future__ import annotations

import json
import sys
from datetime import datetime, timezone

from time_mcp_server.state import load_state, save_state


def _format_seconds(s: int) -> str:
    """Compact human format: 4m, 2h, 1d. Always positive (caller passes |delta|)."""
    if s < 60:
        return f"{s}s"
    if s < 3600:
        return f"{s // 60}m"
    if s < 86400:
        return f"{s // 3600}h"
    return f"{s // 86400}d"


def _label_part(label) -> str:
    return f" '{label}'" if label else ""


def collect_notifications(state: dict, now: datetime) -> tuple[list[str], dict]:
    """Find newly-expired timers + fired alarms; return notification lines and
    a mutated state dict (records updated with ``notified_at``).

    A record qualifies if:
      - cancelled_at is null/missing, AND
      - notified_at is null/missing, AND
      - the threshold time has passed (expires_at for timers, fires_at for alarms).
    """
    lines: list[str] = []
    now_iso = now.isoformat()

    for timer_id, record in state.get("timers", {}).items():
        if record.get("cancelled_at") or record.get("notified_at"):
            continue
        expires_at = datetime.fromisoformat(record["expires_at"])
        if now < expires_at:
            continue
        ago = _format_seconds(int((now - expires_at).total_seconds()))
        lines.append(
            f"\U0001f514 Timer{_label_part(record.get('label'))} ({timer_id}) expired {ago} ago"
        )
        record["notified_at"] = now_iso

    for alarm_id, record in state.get("alarms", {}).items():
        if record.get("cancelled_at") or record.get("notified_at"):
            continue
        fires_at = datetime.fromisoformat(record["fires_at"])
        if now < fires_at:
            continue
        ago = _format_seconds(int((now - fires_at).total_seconds()))
        lines.append(
            f"\U0001f514 Alarm{_label_part(record.get('label'))} ({alarm_id}) fired {ago} ago"
        )
        record["notified_at"] = now_iso

    return lines, state


def main() -> int:
    """Hook entry point. Always returns 0 — never break a user prompt."""
    # Drain the JSON Claude Code sends on stdin; we don't use it but shouldn't
    # leave the pipe unread.
    try:
        sys.stdin.read()
    except Exception:
        pass

    try:
        state = load_state()
    except Exception:
        return 0

    lines, mutated = collect_notifications(state, datetime.now(timezone.utc))
    if not lines:
        return 0

    try:
        save_state(mutated)
    except Exception:
        # If the save fails, still emit notifications this turn; we'll
        # duplicate next turn rather than silently drop.
        pass

    payload = {
        "hookSpecificOutput": {
            "hookEventName": "UserPromptSubmit",
            "additionalContext": "\n".join(lines),
        }
    }
    print(json.dumps(payload))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: src/notify-hook.ts
#!/usr/bin/env node
import { fileURLToPath } from "node:url";
import { DateTime } from "luxon";
import { loadState, saveState } from "./state.js";
import type { State } from "./state.js";

export function formatSeconds(s: number): string {
  if (s < 60) return `${s}s`;
  if (s < 3600) return `${Math.floor(s / 60)}m`;
  if (s < 86400) return `${Math.floor(s / 3600)}h`;
  return `${Math.floor(s / 86400)}d`;
}

function labelPart(label: string | null | undefined): string {
  return label ? ` '${label}'` : "";
}

export function collectNotifications(state: State, now: DateTime): { lines: string[]; state: State } {
  const lines: string[] = [];
  const nowIso = now.toISO() ?? "";

  for (const [timerId, record] of Object.entries(state.timers)) {
    if (record.cancelled_at || record.notified_at) continue;
    const expiresAt = DateTime.fromISO(record.expires_at);
    if (now < expiresAt) continue;
    const ago = formatSeconds(Math.floor(now.diff(expiresAt, "seconds").seconds));
    lines.push(`\u{1F514} Timer${labelPart(record.label)} (${timerId}) expired ${ago} ago`);
    record.notified_at = nowIso;
  }

  for (const [alarmId, record] of Object.entries(state.alarms)) {
    if (record.cancelled_at || record.notified_at) continue;
    const firesAt = DateTime.fromISO(record.fires_at);
    if (now < firesAt) continue;
    const ago = formatSeconds(Math.floor(now.diff(firesAt, "seconds").seconds));
    lines.push(`\u{1F514} Alarm${labelPart(record.label)} (${alarmId}) fired ${ago} ago`);
    record.notified_at = nowIso;
  }

  return { lines, state };
}

async function readStdin(): Promise<void> {
  return new Promise((resolve) => {
    process.stdin.on("data", () => {});
    process.stdin.on("end", () => resolve());
    process.stdin.on("error", () => resolve());
    // If nothing is piped in, resolve quickly.
    setTimeout(() => resolve(), 50);
  });
}

async function main(): Promise<number> {
  try { await readStdin(); } catch { /* ignore */ }

  let state: State;
  try {
    state = await loadState();
  } catch {
    return 0;
  }

  const { lines, state: mutated } = collectNotifications(state, DateTime.utc());
  if (lines.length === 0) return 0;

  try {
    await saveState(mutated);
  } catch {
    // If save fails, still emit notifications — better to duplicate next turn than silently drop.
  }

  const payload = {
    hookSpecificOutput: {
      hookEventName: "UserPromptSubmit",
      additionalContext: lines.join("\n"),
    },
  };
  process.stdout.write(JSON.stringify(payload) + "\n");
  return 0;
}

// Only run as CLI entry point, not when imported by tests or other modules.
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  main().then((code) => process.exit(code)).catch(() => process.exit(0));
}
